openmdao/func_api.py

```diff
@@ -0,0 +1,826 @@
+
+        notfound = []
+        for oname, ometa in self._outputs.items():
+            for n, meta in outlist:
+                if n == oname:
+                    if meta is not ometa:
+                        meta.update(ometa)
+                    break
+            else:  # didn't find oname
+                notfound.append(oname)
+
+        if notfound:  # try to fill in the unnamed slots with user-supplied output data
+            inones = [i for i, (n, m) in enumerate(outlist) if n is None]  # indices with no name
+            if len(notfound) != len(inones):
+                raise RuntimeError(f"There must be an unnamed return value for every unmatched "
+                                   f"output name {notfound} but only found {len(inones)}.")
+
+            # number of None return slots equals number of entries not found in outlist
+            for i_olist, name_notfound in zip(inones, notfound):
+                m = self._outputs[name_notfound]
+                _, ret_meta = outlist[i_olist]
+                ret_meta.update(m)
+                outlist[i_olist] = (name_notfound, ret_meta)
+
+        outs = {n: m for n, m in outlist}
+
+        if self._use_jax:
+            self._compute_out_shapes(self._inputs, outs)
+
+        for meta in outs.values():
+            for o in overrides:
+                self._resolve_default(o, meta)
+            if meta['shape'] is not None:
+                meta['shape'] = _shape2tuple(meta['shape'])
+
+        self._outputs = outs
+
+    def _compute_out_shapes(self, ins, outs):
+        """
+        Compute the shapes of outputs based on those of the inputs.
+
+        Parameters
+        ----------
+        ins : dict
+            Dict of input metadata containing input shapes.
+        outs : dict
+            Dict of output metadata that will be updated with shape information.
+        """
+        need_shape = []
+        for name, ometa in outs.items():
+            try:
+                ometa['shape']
+            except KeyError:
+                need_shape.append(name)
+
+        args = []
+        for name, meta in ins.items():
+            if 'is_option' in meta and meta['is_option']:
+                if 'default' in meta:
+                    val = meta['default']
+                elif 'values' in meta:
+                    val = meta['values'][0]
+                else:
+                    val = None
+                args.append(val)
+                continue
+            if meta['val'] is not None:
+                args.append(meta['val'])
+            else:
+                try:
+                    shp = meta['shape']
+                except KeyError:
+                    raise RuntimeError(f"Can't determine shape of input '{name}'.")
+                if jax is not None:
+                    args.append(jax.ShapedArray(_shape2tuple(shp), dtype=np.float64))
+
+        # compute shapes as a check against annotated value (if any)
+        if jax is not None:
```

Comment:
Why isn't this `self._use_jax` instead? Seems like the user might have jax imported, but they aren't using it and haven't specified it as the declare_partials method.  I guess it just seems weird that importing jax into your local env changes the behavior of this component.  Even though that is a major motivation for this component, there might be a reason it's not being used at a time.

Reply:
This block uses jax only to compute output shapes automatically, not to compute derivatives or anything, so the reasoning for doing it this way is that if jax is available we should always use it as a check on the output shapes.  Unless we're worried about performance or something due to calling jax.make_jaxpr, I don't see why we wouldn't just always do this, assuming jax is installed.

Reply:
I guess one reason might be: they are doing something weird in this function that jax doesn't like, but they imported it for their other functions.

Reply:
Actually I just noticed that _compute_out_shapes doesn't even get called unless self._use_jax is True, so disregard my earlier comment.